src/gc/proxy.d

```diff
@@ -28,32 +28,67 @@ private
     extern (C) void thread_init();
     extern (C) void thread_term();
 
+    import core.internal.spinlock;
+    static SpinLock instanceLock;
+
     __gshared GC instance;
     __gshared GC proxiedGC; // used to iterate roots of Windows DLLs
-
 }
 
+alias GCInitNoThrow = void function() nothrow @nogc;
 
 extern (C)
 {
 
     void gc_init()
     {
-        config.initialize();
-        ManualGC.initialize(instance);
-        ConservativeGC.initialize(instance);
-        if (instance is null)
-        {
-            import core.stdc.stdio : fprintf, stderr;
-            import core.stdc.stdlib : exit;
+        import core.atomic : atomicLoad, atomicStore;
+
+        auto pinstance = cast(shared(GC)*) &instance;
 
-            fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
-            exit(1);
+        if (atomicLoad(*pinstance) is null)
+        {
+            instanceLock.lock();
+            scope(exit) instanceLock.unlock();
+            {
+                if (atomicLoad(*pinstance) is null)
+                {
+                    GC auxInstance;
+                    config.initialize();
+                    ManualGC.initialize(auxInstance);
+                    ConservativeGC.initialize(auxInstance);
+
+                    if (auxInstance is null)
+                    {
+                        import core.stdc.stdio : fprintf, stderr;
+                        import core.stdc.stdlib : exit;
+
+                        fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
+                        instanceLock.unlock();
+                        exit(1);
+                    }
+
+                    atomicStore(*pinstance, cast(shared GC)auxInstance);
+
+                    // NOTE: The GC must initialize the thread library
+                    //       before its first collection.
+                    thread_init();
+                }
+            }
         }
+    }
 
-        // NOTE: The GC must initialize the thread library
-        //       before its first collection.
-        thread_init();
+    void gc_init_nothrow() nothrow
+    {
+        try
```

Comment:
Rather than incurring the overrhead of a try/catch for every `gc_init_nothrow`, this should probably go inside the double checked lock.  That way you only have to pay the price of a try/catch once, when you initialize the GC.

Reply:
Indeed, use sth. similar to [`initOnce`](https://github.com/dlang/phobos/blob/da00e8f3908d221008d7c550bfed97ac9628afb2/std/concurrency.d#L2569-L2589).
Instead of a separate `flag` variable, you could directly load/store the shared `instance` variable. Also seems like the spinLock would be fine to be used as mutex.

Reply:
I think the purpose here was just to trick `nothrow`.
A shorter idiom for that is the following.
```d
void foo() nothrow
{
    scope (failure) assert(0, "unexpected exception");
    stmt; // not yet nothrow annotated
}
```

Reply:
From what I understand, `try/catch` and `scope(failure)` incur some amount of overhead.  By moving this inside the lock we only have to incur it once.  I still recommend doing so.

Reply:
Based on @MartinNowak's suggestions, `initOnce` looks like it could work here so long as you use a mutex and wrap the `init` argument in a tryCatch.

Reply:
dmd is fairly bad at optimizing initOnce, so doing it manually seems fine here.